dlc-messages/src/lib.rs

```diff
@@ -0,0 +1,624 @@
+//! Data structure and functions related to peer communication.
+
+extern crate bitcoin;
+#[macro_use]
+extern crate lightning;
+extern crate dlc;
+extern crate secp256k1;
+extern crate unicode_normalization;
+
+#[cfg(any(test, feature = "serde"))]
+extern crate serde;
+
+#[cfg(test)]
+extern crate serde_json;
+
+use bitcoin::hashes::*;
+use bitcoin::{consensus::Decodable, hash_types::Txid, OutPoint, Script, Transaction};
+use dlc::TxInputInfo;
+use lightning::ln::msgs::DecodeError;
+use lightning::ln::wire::{write, Encode, MessageType};
+use lightning::util::ser::{BigSize, Readable, Writeable, Writer};
+use secp256k1::ecdsa_adaptor::{AdaptorProof, AdaptorSignature};
+use secp256k1::schnorrsig::PublicKey as SchnorrPublicKey;
+use secp256k1::{constants, PublicKey, Signature};
+use unicode_normalization::UnicodeNormalization;
+
+const CONTRACT_INFO_TYPE: u64 = 42768;
+const ORACLE_INFO_TYPE: u64 = 42770;
+const FUNDING_INPUT_TYPE: u64 = 42772;
+const CET_ADAPTOR_SIGNATURES_TYPE: u64 = 42774;
+const FUNDING_SIGNATURES_TYPE: u64 = 42776;
+
+/// Represents a single outcome of a DLC contract and the associated offer party
+/// payout.
+#[derive(Clone, PartialEq)]
+#[cfg_attr(
+    any(test, feature = "serde"),
+    derive(serde::Deserialize, serde::Serialize),
+    serde(rename_all = "camelCase")
+)]
+pub struct ContractOutcome {
+    pub outcome: String,
+    pub local_payout: u64,
+}
+
+impl Writeable for ContractOutcome {
+    fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+        let outcome = &self.outcome;
+        let nfced = outcome.nfc().collect::<String>();
+        let len = BigSize(nfced.len() as u64);
+        len.write(writer)?;
+        let bytes = nfced.as_bytes();
```

Comment:
nit: Might be nice to have factored out utility functions for serializing and deserializing `String`s? You'll need this more often in the future when you add oracle TLVs

Reply:
Done